include/deal.II/multigrid/mg_transfer.templates.h

```diff
@@ -172,24 +175,40 @@ MGTransferPrebuilt<VECTOR>::copy_to_mg (
 {
   reinit_vector(mg_dof_handler, component_to_block_map, dst);
   bool first = true;
+#ifdef DEBUG_OUTPUT
+  std::cout << "copy_to_mg src " << src.l2_norm() << std::endl;
+  MPI_Barrier(MPI_COMM_WORLD);
+#endif
   for (unsigned int level=mg_dof_handler.get_tria().n_global_levels(); level != 0;)
     {
       --level;
       VECTOR &dst_level = dst[level];
 
+#ifdef DEBUG_OUTPUT
+      MPI_Barrier(MPI_COMM_WORLD);
+#endif
+
       typedef std::vector<std::pair<types::global_dof_index, types::global_dof_index> >::const_iterator IT;
       for (IT i= copy_indices[level].begin();
            i != copy_indices[level].end(); ++i)
         dst_level(i->second) = src(i->first);
 
-      // for (IT i= copy_indices_to_me[level].begin();
-      //      i != copy_indices_to_me[level].end(); ++i)
-      //   dst_level(i->second) = src(i->first);
+      for (IT i= copy_indices_global_mine[level].begin();
+           i != copy_indices_global_mine[level].end(); ++i)
+        dst_level(i->second) = src(i->first);
```

Comment:
Yes. When copying from/to a multigrid vector, I can have four cases for each dof: level and global is locally owned, one but not the other, and neither is. I want the copy operation to happen exactly once (because it actually might do an add). The way it is currently done is that I am reading from locally owned dofs and writing into dofs that might belong to somebody else. (note that to_mg() uses copy_indices_global_mine and from_mg() uses copy_indices_level_mine). Otherwise I would need to import ghost values.

To make this work with p::d::Vector we would need to import the ghost values and write locally (turning the logic around).

Also note that one of the nasty bugs was that processors would not agree on the local/global_mine data structures because their view of the world can be different. I finally fixed this by doing an MPI communication.